llvm_mode/afl-llvm-pass.so.cc

```diff
@@ -87,88 +94,75 @@ bool AFLCoverage::runOnModule(Module &M) {
+  bool write_to_file = false;
+  std::error_code llvm_of_error; 
+  const char* output_file_name = std::getenv("BB_LOGFILE_NAME");
 
+  if (output_file_name != nullptr) {
+    location_file.open(output_file_name, std::ios::trunc);
+    write_to_file = true;
+  }
+  
   for (auto &F : M)
     for (auto &BB : F) {
 
+      assert(block_counter < MAP_SIZE && "counter is too large");
       BasicBlock::iterator IP = BB.getFirstInsertionPt();
       IRBuilder<> IRB(&(*IP));
 
-      if (AFL_R(100) >= inst_ratio) continue;
-
-      /* Make up cur_loc */
-
-      unsigned int cur_loc = AFL_R(MAP_SIZE);
-
-      ConstantInt *CurLoc = ConstantInt::get(Int32Ty, cur_loc);
-
-      /* Load prev_loc */
+      /* Make up cur_loc based on block counter */
 
-      LoadInst *PrevLoc = IRB.CreateLoad(AFLPrevLoc);
-      PrevLoc->setMetadata(M.getMDKindID("nosanitize"), MDNode::get(C, None));
-      Value *PrevLocCasted = IRB.CreateZExt(PrevLoc, IRB.getInt32Ty());
+      ConstantInt *CurBB = ConstantInt::get(Int64Ty, block_counter);
 
       /* Load SHM pointer */
 
-      LoadInst *MapPtr = IRB.CreateLoad(AFLMapPtr);
+      LoadInst *MapPtr = IRB.CreateLoad(BBCoverageMapPtr);
       MapPtr->setMetadata(M.getMDKindID("nosanitize"), MDNode::get(C, None));
-      Value *MapPtrIdx =
-          IRB.CreateGEP(MapPtr, IRB.CreateXor(PrevLocCasted, CurLoc));
+      Value *MapPtrIdx = IRB.CreateGEP(MapPtr,  CurBB);
 
       /* Update bitmap */
 
       LoadInst *Counter = IRB.CreateLoad(MapPtrIdx);
       Counter->setMetadata(M.getMDKindID("nosanitize"), MDNode::get(C, None));
-      Value *Incr = IRB.CreateAdd(Counter, ConstantInt::get(Int8Ty, 1));
+      Value *Incr = IRB.CreateAdd(Counter, ConstantInt::get(Int64Ty, 1));
       IRB.CreateStore(Incr, MapPtrIdx)
           ->setMetadata(M.getMDKindID("nosanitize"), MDNode::get(C, None));
 
-      /* Set prev_loc to cur_loc >> 1 */
+      if (write_to_file){
+        
+        /* Make a raw_string_ostream to save the content of a block into a string */
 
-      StoreInst *Store =
-          IRB.CreateStore(ConstantInt::get(Int32Ty, cur_loc >> 1), AFLPrevLoc);
-      Store->setMetadata(M.getMDKindID("nosanitize"), MDNode::get(C, None));
+        std::string block_string;
+        llvm::raw_string_ostream block_string_stream(block_string);
+        block_string_stream << BB;
+        SmallString<32> block_md5 = getmd5(block_string_stream.str());
+        location_file << M.getSourceFileName() << ":" << F.getName().str() << ":" << block_md5.c_str() << ":" << block_counter << "\n";
 
-      inst_blocks++;
+      }
+            
+      block_counter++;
 
     }
 
   /* Say something nice. */
 
   if (!be_quiet) {
 
-    if (!inst_blocks) WARNF("No instrumentation targets found.");
-    else OKF("Instrumented %u locations (%s mode, ratio %u%%).",
-             inst_blocks, getenv("AFL_HARDEN") ? "hardened" :
+    if (!block_counter) WARNF("No instrumentation targets found.");
+    else OKF("Instrumented %u locations (%s mode).",
```

Comment:
I suspect you'll get an compiler warning here, because %u is for unsigned int, but we're passing an u64, you should use PRIu64